File: src/algorithm_java/leetcode/Problem200.java
package algorithm.leetcode;

public class Problem200 {

    public static boolean[][] used;

    public static void searchIsland(char[][] grid, int i, int j) {
        if (grid[i][j] == '0')
            return;
        else if (grid[i][j] == '1') {
            used[i][j] = true;
            if (i - 1 >= 0 && !used[i - 1][j])
                searchIsland(grid, i - 1, j);
            if (j - 1 >= 0 && !used[i][j - 1])
                searchIsland(grid, i, j - 1);
            if (i + 1 < grid.length && !used[i + 1][j])
                searchIsland(grid, i + 1, j);
            if (j + 1 < grid[0].length && !used[i][j + 1])
                searchIsland(grid, i, j + 1);
        }
    }

    public static int numIslands(char[][] grid) {
        if (grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        int count = 0;
        used = new boolean[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (!used[i][j] && grid[i][j] == '1') {
                    searchIsland(grid, i, j);
                    count++;
                }
            }
        }
        return count;
    }
}

File: src/algorithm_cpp/problem200.cpp
#include<vector>

using namespace std;

class Solution {
public:
    void visit(vector<vector<char>>& grid, int i, int j) {
	if (grid[i][j] == '0') return;
	grid[i][j] = '0';
	if (i - 1 >= 0) visit(grid, i - 1, j);
	if (i + 1 < grid.size()) visit(grid, i + 1, j);
	if (j - 1 >= 0) visit(grid, i, j - 1);
	if (j + 1 < grid[0].size()) visit(grid, i, j + 1);
    }

    int numIslands(vector<vector<char>>& grid) {
	int count = 0;
	for (int i = 0; i < grid.size(); i++) {
	    for (int j = 0; j < grid[0].size(); j++) {
		if (grid[i][j] == '1') count++;
		visit(grid, i, j);
	    }
	}
	return count;
    }
};

int main() {
    return 0;
}
